File: frontend/src/types/ai-analysis.ts
/**
 * AI Analysis TypeScript interfaces
 */

import { BaseEntity, PaginatedResponse, Timestamp, UUID } from './base';

// AI Analysis Request Types
export type AnalysisType = 
  | 'fundamental' 
  | 'technical' 
  | 'sentiment' 
  | 'risk_assessment'
  | 'price_prediction' 
  | 'earnings_forecast' 
  | 'peer_comparison'
  | 'market_outlook' 
  | 'investment_recommendation';

export type AnalysisLanguage = 'ja' | 'en';

export interface AIAnalysisRequest {
  ticker: string;
  analysis_type: AnalysisType;
  parameters?: Record<string, any>;
  language: AnalysisLanguage;
}

export interface BulkAnalysisRequest {
  tickers: string[];
  analysis_type: AnalysisType;
  parameters?: Record<string, any>;
  language: AnalysisLanguage;
}

// AI Analysis Response Types
export type AnalysisStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface AIAnalysis extends BaseEntity {
  ticker: string;
  analysis_type: AnalysisType;
  status: AnalysisStatus;
  language: AnalysisLanguage;
  user_id: UUID;
  result?: Record<string, any>;
  error_message?: string;
  processing_time_seconds?: number;
  tokens_used?: number;
  cost_usd?: number;
}

export interface AIAnalysisCreate {
  ticker: string;
  analysis_type: AnalysisType;
  status: AnalysisStatus;
  language: AnalysisLanguage;
  user_id: UUID;
  parameters?: Record<string, any>;
}

// Specific Analysis Result Types
export interface FundamentalAnalysisResult {
  overall_score: number; // 0-100
  financial_health: Record<string, any>;
  valuation: Record<string, any>;
  growth_prospects: Record<string, any>;
  competitive_position: Record<string, any>;
  risks: string[];
  opportunities: string[];
  summary: string;
  recommendation: string;
  target_price?: number;
  confidence_level: number; // 0-1
}

export type TechnicalSignal = 'strong_buy' | 'buy' | 'hold' | 'sell' | 'strong_sell';

export interface TechnicalAnalysisResult {
  overall_signal: TechnicalSignal;
  trend_analysis: Record<string, any>;
  support_resistance: Record<string, any>;
  indicators: Record<string, any>;
  chart_patterns: string[];
  volume_analysis: Record<string, any>;
  momentum: Record<string, any>;
  summary: string;
  short_term_outlook: string;
  medium_term_outlook: string;
  key_levels: Record<string, number>;
}

export type SentimentLevel = 'very_positive' | 'positive' | 'neutral' | 'negative' | 'very_negative';

export interface SentimentAnalysisResult {
  overall_sentiment: SentimentLevel;
  sentiment_score: number; // -1 to 1
  news_sentiment: Record<string, any>;
  social_sentiment: Record<string, any>;
  analyst_sentiment: Record<string, any>;
  sentiment_trends: Array<Record<string, any>>;
  key_themes: string[];
  sentiment_drivers: string[];
  summary: string;
}

export type RiskLevel = 'very_low' | 'low' | 'moderate' | 'high' | 'very_high';

export interface RiskAssessmentResult {
  overall_risk_level: RiskLevel;
  risk_score: number; // 0-100
  financial_risks: Array<Record<string, any>>;
  market_risks: Array<Record<string, any>>;
  operational_risks: Array<Record<string, any>>;
  regulatory_risks: Array<Record<string, any>>;
  volatility_analysis: Record<string, any>;
  correlation_analysis: Record<string, any>;
  risk_mitigation: string[];
  summary: string;
}

export type PredictionHorizon = '1_week' | '1_month' | '3_months' | '6_months' | '1_year';

export interface PricePredictionResult {
  prediction_horizon: PredictionHorizon;
  predicted_price: number;
  current_price: number;
  price_change_percent: number;
  confidence_interval: Record<string, number>;
  prediction_factors: string[];
  model_accuracy?: number;
  scenarios: Record<string, Record<string, number>>;
  risks_to_prediction: string[];
  summary: string;
}

// Analysis History Types
export interface AnalysisHistory {
  user_id: UUID;
  analyses: AIAnalysis[];
  total_analyses: number;
  analyses_this_month: number;
  favorite_analysis_types: string[];
  total_cost_usd: number;
}

// Analysis Queue Types
export interface AnalysisQueueStatus {
  queue_position: number;
  estimated_wait_time: number; // seconds
  total_queue_size: number;
  processing_capacity: number;
}

// Analysis Comparison Types
export interface AnalysisComparison {
  tickers: string[];
  analysis_type: AnalysisType;
  comparison_metrics: Record<string, any>;
  rankings: Array<Record<string, any>>;
  summary: string;
  created_at: Timestamp;
}

// Analysis Export Types
export type AnalysisExportFormat = 'pdf' | 'excel' | 'csv' | 'json';

export interface AnalysisExportRequest {
  analysis_ids: UUID[];
  format: AnalysisExportFormat;
  include_charts: boolean;
  language: AnalysisLanguage;
}

export type ExportStatus = 'pending' | 'processing' | 'completed' | 'failed' | 'expired';

export interface AnalysisExportResponse {
  export_id: UUID;
  download_url?: string;
  status: ExportStatus;
  created_at: Timestamp;
  expires_at?: Timestamp;
}

// Analysis Statistics Types
export interface AnalysisStatistics {
  total_analyses: number;
  analyses_by_type: Record<string, number>;
  analyses_by_language: Record<string, number>;
  average_processing_time: number;
  success_rate: number;
  total_cost_usd: number;
  most_analyzed_stocks: Array<Record<string, any>>;
  peak_usage_hours: number[];
  user_satisfaction_score?: number;
}

// Analysis Configuration Types
export interface AnalysisConfig {
  analysis_type: AnalysisType;
  parameters: {
    time_horizon?: PredictionHorizon;
    include_technical_indicators?: boolean;
    include_fundamental_data?: boolean;
    include_news_sentiment?: boolean;
    include_peer_comparison?: boolean;
    risk_tolerance?: 'conservative' | 'moderate' | 'aggressive';
    investment_style?: 'value' | 'growth' | 'momentum' | 'dividend';
    market_conditions?: 'bull' | 'bear' | 'sideways';
  };
  output_preferences: {
    language: AnalysisLanguage;
    detail_level: 'summary' | 'detailed' | 'comprehensive';
    include_charts: boolean;
    include_recommendations: boolean;
    format: 'text' | 'structured' | 'bullet_points';
  };
}

// Analysis Template Types
export interface AnalysisTemplate {
  id: UUID;
  name: string;
  description: string;
  analysis_type: AnalysisType;
  config: AnalysisConfig;
  is_public: boolean;
  created_by: UUID;
  usage_count: number;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface AnalysisTemplateCreate {
  name: string;
  description: string;
  analysis_type: AnalysisType;
  config: AnalysisConfig;
  is_public?: boolean;
}

// Analysis Feedback Types
export interface AnalysisFeedback {
  id: UUID;
  analysis_id: UUID;
  user_id: UUID;
  rating: number; // 1-5 scale
  feedback_text?: string;
  helpful_aspects: string[];
  improvement_suggestions: string[];
  would_recommend: boolean;
  created_at: Timestamp;
}

export interface AnalysisFeedbackCreate {
  analysis_id: UUID;
  rating: number;
  feedback_text?: string;
  helpful_aspects: string[];
  improvement_suggestions: string[];
  would_recommend: boolean;
}

// Analysis Sharing Types
export interface AnalysisShare {
  id: UUID;
  analysis_id: UUID;
  shared_by: UUID;
  share_token: string;
  is_public: boolean;
  expires_at?: Timestamp;
  view_count: number;
  created_at: Timestamp;
}

export interface AnalysisShareCreate {
  analysis_id: UUID;
  is_public: boolean;
  expires_at?: Timestamp;
}

// Analysis State Types (for Redux/Context)
export interface AIAnalysisState {
  // Current analyses
  analyses: {
    [ticker: string]: {
      [analysisType: string]: AIAnalysis;
    };
  };
  
  // Analysis history
  history: AIAnalysis[];
  historyPagination: {
    page: number;
    total_pages: number;
    total_items: number;
  };
  
  // Queue status
  queueStatus: AnalysisQueueStatus | null;
  
  // Templates
  templates: AnalysisTemplate[];
  
  // Current analysis request
  currentRequest: AIAnalysisRequest | null;
  
  // Loading states
  isLoadingAnalysis: boolean;
  isLoadingHistory: boolean;
  isLoadingTemplates: boolean;
  isSubmittingRequest: boolean;
  
  // Error states
  analysisError?: string;
  historyError?: string;
  templatesError?: string;
  requestError?: string;
  
  // UI state
  selectedAnalysisType: AnalysisType;
  selectedLanguage: AnalysisLanguage;
  analysisConfig: AnalysisConfig;
  
  // Statistics
  statistics: AnalysisStatistics | null;
  
  // Last updated timestamps
  lastUpdated: {
    analyses?: Timestamp;
    history?: Timestamp;
    templates?: Timestamp;
    statistics?: Timestamp;
  };
}

// Analysis Form Types
export interface AnalysisFormData {
  ticker: string;
  analysis_type: AnalysisType;
  language: AnalysisLanguage;
  parameters: Record<string, any>;
  save_as_template: boolean;
  template_name?: string;
  template_description?: string;
}

export interface BulkAnalysisFormData {
  tickers: string[];
  analysis_type: AnalysisType;
  language: AnalysisLanguage;
  parameters: Record<string, any>;
  batch_name?: string;
}

// Analysis Display Types
export interface AnalysisDisplayConfig {
  show_summary: boolean;
  show_details: boolean;
  show_charts: boolean;
  show_recommendations: boolean;
  show_risks: boolean;
  show_metadata: boolean;
  compact_view: boolean;
}

export interface AnalysisCard {
  analysis: AIAnalysis;
  displayConfig: AnalysisDisplayConfig;
  onShare?: (analysis: AIAnalysis) => void;
  onExport?: (analysis: AIAnalysis) => void;
  onFeedback?: (analysis: AIAnalysis) => void;
  onDelete?: (analysis: AIAnalysis) => void;
}

// Paginated AI Analysis Response
export type PaginatedAIAnalysisResponse = PaginatedResponse<AIAnalysis>;

File: backend/app/schemas/ai_analysis.py
"""AI analysis Pydantic schemas."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.schemas.base import BaseSchema, PaginatedResponse, TimestampSchema, UUIDSchema


# AI Analysis Request Schemas
class AIAnalysisRequest(BaseModel):
    """AI analysis request schema."""

    ticker: str = Field(..., max_length=10, description="Stock ticker to analyze")
    analysis_type: str = Field(..., description="Type of analysis requested")
    parameters: Optional[Dict[str, Any]] = Field(
        None, description="Analysis parameters"
    )
    language: str = Field("ja", description="Response language (ja, en)")

    @field_validator("analysis_type")
    @classmethod
    def validate_analysis_type(cls, v):
        """Validate analysis type."""
        valid_types = [
            "fundamental",
            "technical",
            "sentiment",
            "risk_assessment",
            "price_prediction",
            "earnings_forecast",
            "peer_comparison",
            "market_outlook",
            "investment_recommendation",
        ]
        if v not in valid_types:
            raise ValueError(f'Analysis type must be one of: {", ".join(valid_types)}')
        return v

    @field_validator("language")
    @classmethod
    def validate_language(cls, v):
        """Validate language code."""
        if v not in ["ja", "en"]:
            raise ValueError("Language must be ja or en")
        return v


class BulkAnalysisRequest(BaseModel):
    """Bulk AI analysis request schema."""

    tickers: List[str] = Field(
        ..., min_items=1, max_items=50, description="Stock tickers to analyze"
    )
    analysis_type: str = Field(..., description="Type of analysis requested")
    parameters: Optional[Dict[str, Any]] = Field(
        None, description="Analysis parameters"
    )
    language: str = Field("ja", description="Response language (ja, en)")

    @field_validator("tickers")
    @classmethod
    def validate_tickers(cls, v):
        """Validate ticker list."""
        if len(set(v)) != len(v):
            raise ValueError("Duplicate tickers are not allowed")
        return v

    @field_validator("analysis_type")
    @classmethod
    def validate_analysis_type(cls, v):
        """Validate analysis type."""
        valid_types = [
            "fundamental",
            "technical",
            "sentiment",
            "risk_assessment",
            "price_prediction",
            "earnings_forecast",
            "peer_comparison",
            "market_outlook",
            "investment_recommendation",
        ]
        if v not in valid_types:
            raise ValueError(f'Analysis type must be one of: {", ".join(valid_types)}')
        return v


# AI Analysis Response Schemas
class AIAnalysisBase(BaseModel):
    """Base AI analysis schema."""

    ticker: str = Field(..., max_length=10, description="Stock ticker")
    analysis_type: str = Field(..., description="Type of analysis")
    status: str = Field(..., description="Analysis status")
    language: str = Field(..., description="Response language")
    user_id: UUID = Field(..., description="User who requested analysis")

    @field_validator("status")
    @classmethod
    def validate_status(cls, v):
        """Validate analysis status."""
        if v not in ["pending", "processing", "completed", "failed"]:
            raise ValueError(
                "Status must be one of: pending, processing, completed, failed"
            )
        return v


class AIAnalysisCreate(AIAnalysisBase):
    """AI analysis creation schema."""

    parameters: Optional[Dict[str, Any]] = Field(
        None, description="Analysis parameters"
    )


class AIAnalysis(UUIDSchema, AIAnalysisBase, TimestampSchema):
    """AI analysis response schema."""

    result: Optional[Dict[str, Any]] = Field(None, description="Analysis result")
    error_message: Optional[str] = Field(None, description="Error message if failed")
    processing_time_seconds: Optional[float] = Field(
        None, description="Processing time in seconds"
    )
    tokens_used: Optional[int] = Field(None, description="AI tokens consumed")
    cost_usd: Optional[float] = Field(None, description="Analysis cost in USD")


# Specific Analysis Result Schemas
class FundamentalAnalysisResult(BaseModel):
    """Fundamental analysis result schema."""

    overall_score: float = Field(
        ..., ge=0, le=100, description="Overall fundamental score"
    )
    financial_health: Dict[str, Any] = Field(
        ..., description="Financial health metrics"
    )
    valuation: Dict[str, Any] = Field(..., description="Valuation analysis")
    growth_prospects: Dict[str, Any] = Field(..., description="Growth analysis")
    competitive_position: Dict[str, Any] = Field(
        ..., description="Competitive analysis"
    )
    risks: List[str] = Field(..., description="Identified risks")
    opportunities: List[str] = Field(..., description="Identified opportunities")
    summary: str = Field(..., description="Analysis summary")
    recommendation: str = Field(..., description="Investment recommendation")
    target_price: Optional[float] = Field(None, description="Target price")
    confidence_level: float = Field(..., ge=0, le=1, description="Analysis confidence")


class TechnicalAnalysisResult(BaseModel):
    """Technical analysis result schema."""

    overall_signal: str = Field(..., description="Overall technical signal")
    trend_analysis: Dict[str, Any] = Field(..., description="Trend analysis")
    support_resistance: Dict[str, Any] = Field(
        ..., description="Support/resistance levels"
    )
    indicators: Dict[str, Any] = Field(..., description="Technical indicators")
    chart_patterns: List[str] = Field(..., description="Identified chart patterns")
    volume_analysis: Dict[str, Any] = Field(..., description="Volume analysis")
    momentum: Dict[str, Any] = Field(..., description="Momentum indicators")
    summary: str = Field(..., description="Technical summary")
    short_term_outlook: str = Field(..., description="Short-term outlook")
    medium_term_outlook: str = Field(..., description="Medium-term outlook")
    key_levels: Dict[str, float] = Field(..., description="Key price levels")

    @field_validator("overall_signal")
    @classmethod
    def validate_signal(cls, v):
        """Validate technical signal."""
        if v not in ["strong_buy", "buy", "hold", "sell", "strong_sell"]:
            raise ValueError(
                "Signal must be one of: strong_buy, buy, hold, sell, strong_sell"
            )
        return v


class SentimentAnalysisResult(BaseModel):
    """Sentiment analysis result schema."""

    overall_sentiment: str = Field(..., description="Overall sentiment")
    sentiment_score: float = Field(
        ..., ge=-1, le=1, description="Sentiment score (-1 to 1)"
    )
    news_sentiment: Dict[str, Any] = Field(..., description="News sentiment analysis")
    social_sentiment: Dict[str, Any] = Field(..., description="Social media sentiment")
    analyst_sentiment: Dict[str, Any] = Field(..., description="Analyst sentiment")
    sentiment_trends: List[Dict[str, Any]] = Field(
        ..., description="Sentiment trends over time"
    )
    key_themes: List[str] = Field(..., description="Key sentiment themes")
    sentiment_drivers: List[str] = Field(..., description="Main sentiment drivers")
    summary: str = Field(..., description="Sentiment summary")

    @field_validator("overall_sentiment")
    @classmethod
    def validate_sentiment(cls, v):
        """Validate sentiment."""
        if v not in [
            "very_positive",
            "positive",
            "neutral",
            "negative",
            "very_negative",
        ]:
            raise ValueError(
                "Sentiment must be one of: very_positive, positive, neutral, negative, very_negative"
            )
        return v


class RiskAssessmentResult(BaseModel):
    """Risk assessment result schema."""

    overall_risk_level: str = Field(..., description="Overall risk level")
    risk_score: float = Field(..., ge=0, le=100, description="Risk score (0-100)")
    financial_risks: List[Dict[str, Any]] = Field(..., description="Financial risks")
    market_risks: List[Dict[str, Any]] = Field(..., description="Market risks")
    operational_risks: List[Dict[str, Any]] = Field(
        ..., description="Operational risks"
    )
    regulatory_risks: List[Dict[str, Any]] = Field(..., description="Regulatory risks")
    volatility_analysis: Dict[str, Any] = Field(..., description="Volatility analysis")
    correlation_analysis: Dict[str, Any] = Field(
        ..., description="Correlation analysis"
    )
    risk_mitigation: List[str] = Field(..., description="Risk mitigation strategies")
    summary: str = Field(..., description="Risk assessment summary")

    @field_validator("overall_risk_level")
    @classmethod
    def validate_risk_level(cls, v):
        """Validate risk level."""
        if v not in ["very_low", "low", "moderate", "high", "very_high"]:
            raise ValueError(
                "Risk level must be one of: very_low, low, moderate, high, very_high"
            )
        return v


class PricePredictionResult(BaseModel):
    """Price prediction result schema."""

    prediction_horizon: str = Field(..., description="Prediction time horizon")
    predicted_price: float = Field(..., description="Predicted price")
    current_price: float = Field(..., description="Current price")
    price_change_percent: float = Field(..., description="Predicted price change %")
    confidence_interval: Dict[str, float] = Field(
        ..., description="Confidence intervals"
    )
    prediction_factors: List[str] = Field(..., description="Key prediction factors")
    model_accuracy: Optional[float] = Field(None, description="Model accuracy score")
    scenarios: Dict[str, Dict[str, float]] = Field(
        ..., description="Different scenarios"
    )
    risks_to_prediction: List[str] = Field(..., description="Risks to prediction")
    summary: str = Field(..., description="Prediction summary")

    @field_validator("prediction_horizon")
    @classmethod
    def validate_horizon(cls, v):
        """Validate prediction horizon."""
        if v not in ["1_week", "1_month", "3_months", "6_months", "1_year"]:
            raise ValueError(
                "Horizon must be one of: 1_week, 1_month, 3_months, 6_months, 1_year"
            )
        return v


# Analysis History Schemas
class AnalysisHistory(BaseModel):
    """Analysis history schema."""

    user_id: UUID = Field(..., description="User ID")
    analyses: List[AIAnalysis] = Field(..., description="User's analysis history")
    total_analyses: int = Field(..., description="Total number of analyses")
    analyses_this_month: int = Field(..., description="Analyses this month")
    favorite_analysis_types: List[str] = Field(
        ..., description="Most used analysis types"
    )
    total_cost_usd: float = Field(..., description="Total cost of analyses")


# Analysis Queue Schemas
class AnalysisQueueStatus(BaseModel):
    """Analysis queue status schema."""

    queue_position: int = Field(..., description="Position in queue")
    estimated_wait_time: int = Field(..., description="Estimated wait time in seconds")
    total_queue_size: int = Field(..., description="Total queue size")
    processing_capacity: int = Field(..., description="Current processing capacity")


# Analysis Comparison Schemas
class AnalysisComparison(BaseModel):
    """Analysis comparison schema."""

    tickers: List[str] = Field(..., description="Compared tickers")
    analysis_type: str = Field(..., description="Analysis type")
    comparison_metrics: Dict[str, Any] = Field(..., description="Comparison metrics")
    rankings: List[Dict[str, Any]] = Field(..., description="Ticker rankings")
    summary: str = Field(..., description="Comparison summary")
    created_at: datetime = Field(..., description="Comparison creation time")


# Analysis Export Schemas
class AnalysisExportRequest(BaseModel):
    """Analysis export request schema."""

    analysis_ids: List[UUID] = Field(
        ..., min_items=1, max_items=100, description="Analysis IDs to export"
    )
    format: str = Field("pdf", description="Export format")
    include_charts: bool = Field(True, description="Include charts in export")
    language: str = Field("ja", description="Export language")

    @field_validator("format")
    @classmethod
    def validate_format(cls, v):
        """Validate export format."""
        if v not in ["pdf", "excel", "csv", "json"]:
            raise ValueError("Format must be one of: pdf, excel, csv, json")
        return v


class AnalysisExportResponse(BaseModel):
    """Analysis export response schema."""

    export_id: UUID = Field(..., description="Export job ID")
    download_url: Optional[str] = Field(None, description="Download URL when ready")
    status: str = Field(..., description="Export status")
    created_at: datetime = Field(..., description="Export creation time")
    expires_at: Optional[datetime] = Field(None, description="Download expiration time")

    @field_validator("status")
    @classmethod
    def validate_status(cls, v):
        """Validate export status."""
        if v not in ["pending", "processing", "completed", "failed", "expired"]:
            raise ValueError(
                "Status must be one of: pending, processing, completed, failed, expired"
            )
        return v


# Paginated AI Analysis Responses
class PaginatedAIAnalysisResponse(PaginatedResponse):
    """Paginated AI analysis response."""

    items: List[AIAnalysis]


# Analysis Statistics Schemas
class AnalysisStatistics(BaseModel):
    """Analysis statistics schema."""

    total_analyses: int = Field(..., description="Total analyses performed")
    analyses_by_type: Dict[str, int] = Field(..., description="Analyses by type")
    analyses_by_language: Dict[str, int] = Field(
        ..., description="Analyses by language"
    )
    average_processing_time: float = Field(..., description="Average processing time")
    success_rate: float = Field(..., description="Analysis success rate")
    total_cost_usd: float = Field(..., description="Total cost")
    most_analyzed_stocks: List[Dict[str, Any]] = Field(
        ..., description="Most analyzed stocks"
    )
    peak_usage_hours: List[int] = Field(..., description="Peak usage hours")
    user_satisfaction_score: Optional[float] = Field(
        None, description="User satisfaction score"
    )
